### PR TITLE
Add histogram for E2E pre execution duration

### PR DESCRIPTION
**Pull request by yontyon**
We add histogram (with AsyncRecorder) in the pre execution engine

### CODE REVIEW

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 1]
```diff
@@ -96,7 +96,9 @@ PreProcessor::PreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
                            metricsComponent_.RegisterCounter("preProcReqForwardedByNonPrimaryNotIgnored"),
                            metricsComponent_.RegisterGauge("PreProcInFlyRequestsNum", 0)},
       preExecReqStatusCheckPeriodMilli_(myReplica_.getReplicaConfig().preExecReqStatusCheckTimerMillisec),
-      timers_{timers} {
+      timers_{timers},
+      preExecuteDuration_{histograms_.preExecutionTotal},
+      lastViewNum(myReplica.getCurrentView()) {
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:101]: lastViewNum -> lastViewNum_

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 2]
```diff
@@ -217,6 +219,10 @@ bool PreProcessor::checkClientMsgCorrectness(const ClientPreProcessReqMsgUniqueP
 }
 
 void PreProcessor::updateAggregatorAndDumpMetrics() {
+  if (myReplica_.getCurrentView() != lastViewNum) {
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:222]: Let's clean the map, in any case, periodically. Let's say, every hour.

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 3]
```diff
@@ -472,6 +478,7 @@ void PreProcessor::cancelPreProcessing(NodeIdType clientId) {
       LOG_WARN(logger(), "Pre-processing consensus not reached - cancel request" << KVLOG(reqSeqNum, clientId));
     }
   }
+  if (myReplica_.isCurrentPrimary()) preExecuteDuration_.end(reqSeqNum);
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:481]: reqSeqNum is not unique, go on cid.

File: bftengine/src/preprocessor/PreProcessor.hpp [hunk 4]
```diff
@@ -180,6 +180,8 @@ class PreProcessor {
   const uint64_t preExecReqStatusCheckPeriodMilli_;
   concordUtil::Timers &timers_;
   PreProcessorRecorder histograms_;
+  concord::diagnostics::AsyncTimeRecorderMap<SeqNum> preExecuteDuration_;
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.hpp:183]: Having this, we probably don't need the change in PreProcessorRecorder?

**Review by yuliasherman — CHANGES_REQUESTED** [review]

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 1 above)
Comment (yontyon) [reply bftengine/src/preprocessor/PreProcessor.cpp]: done

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 2 above)
Comment (yontyon) [reply bftengine/src/preprocessor/PreProcessor.cpp]: done

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 3 above)
Comment (yontyon) [reply bftengine/src/preprocessor/PreProcessor.cpp]: done

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 5]
```diff
@@ -472,13 +478,14 @@ void PreProcessor::cancelPreProcessing(NodeIdType clientId) {
   {
     lock_guard<mutex> lock(clientEntry->mutex);
     if (clientEntry->reqProcessingStatePtr) {
+      auto cid = clientEntry->reqProcessingStatePtr->getReqCid();
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:481]: -> auto const &cid

File: bftengine/src/preprocessor/PreProcessor.hpp [hunk 6]
```diff
@@ -163,6 +163,8 @@ class PreProcessor {
   OngoingReqMap ongoingRequests_;  // clientId -> ClientRequestStateSharedPtr
   concordMetrics::Component metricsComponent_;
   std::chrono::seconds metricsLastDumpTime_;
+  std::chrono::steady_clock::time_point lastCleanHistTime_;
+  int64_t clean_hist_interval_ = 3600;  // One hour
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.hpp:167]: -> const

**Review by yuliasherman — APPROVED** [review]
LGTM

**Pull request merged**